feat(components): Carousel and Step components

voronianski opened the pull request:
Addresses proposal https://github.com/sumup/circuit-ui/issues/409.

## Approach and changes

- add generic `Step` component and hook for managing state in step/index based UIs and components 
- add generic `AspectRatio` component for keeping proper image proportions
- add stateful version `Carousel` component powered by `Step` and `AspectRatio` components
- add smaller blocks of the `Carousel` like `Buttons`, `Progress`, `Status`, `Slide`, etc. for composing them in a Custom carousel experience

## Step component

Technically component uses the "function as child" and "prop getter" patterns but it's also possible to use it just as a hook:

```js
import { Step } from '@sumup/circuit-ui';

const config = { ... };

<Step {...config}>
  {({ state, actions, ...propGetters }) => {
    // render your component here
  }}
</Step>
```

vs.

```js
import { useStep } from '@sumup/circuit-ui';

const config = { ... };
const { state, actions, ...propGetters } = useStep(config);
```

Here are the examples below of how `Step` component can be used in a different ways to build index-based UIs.

### Simple image slider

![step-image-slider](https://user-images.githubusercontent.com/974035/65807498-0dd2c300-e18f-11e9-8ae8-0a5b559b4e0b.gif)

```js
import { Step, Button, ButtonGroup, Image } from '@sumup/circuit-ui';

const figureStyles = ({ step, animationDuration }) => css`
  transform: translate3d(${-step * 300}px, 0, 0);
  transition: all ${animationDuration}ms ease-in-out;
`;
const Figure = styled('figure')(figureStyles);

const ImagelSlider = ({ images = [] }) => (
  <Step totalSteps={images.length}>
    {({
      state,
      getNextControlProps,
      getPreviousControlProps,
      getPauseControlProps,
      getPlayControlProps
    }) => (
      <div>
        <Figure
          step={state.step}
          animationDuration={state.animationDuration}
        >
          {images.map((image, i) => (
            <Image
              key={i}
              src={image}
              alt="random pic"
            />
          ))}
        </Figure>
        <ButtonGroup>
          <Button {...getPreviousControlProps()}>Prev</Button>
          <Button
            primary
            {...(state.paused ? getPlayControlProps() : getPauseControlProps())}
          >
            {state.paused ? 'Play' : 'Pause'}
          </Button>
          <Button {...getNextControlProps()}>Next</Button>
        </ButtonGroup>
      </div>
    )}
  </Step>
);
```

### Multistep form

`Step` is suitable not only for image sliders. Any multi-step guides or slideshows can be achieved with it too:

![step-form](https://user-images.githubusercontent.com/974035/65807530-31960900-e18f-11e9-81ea-55ba0f6f017b.gif)
```js
import { Step, Label, Input, Select, Button, ButtonGroup, Heading, ProgressBar } from '@sumup/circuit-ui';

const FormOne = ({ onNextClick }) => (
  <section>
    <Label htmlFor="first">First Name</Label>
    <Input placeholder="John" id="first" />
    <Label htmlFor="second">Second Name</Label>
    <Input placeholder="Doe" id="second" />
    <Button primary onClick={() => onNextClick()}>
      Next
    </Button>
  </section>
);

const FormTwo = ({ onNextClick, onBackClick }) => (
  <section>
    <Label htmlFor="street">Street</Label>
    <Input placeholder="Madison Ave 5" id="street" />
    <Label htmlFor="state">State</Label>
    <Select id="state">
      <option>CA</option>
      <option>TX</option>
      <option>NY</option>
    </Select>
    <Label htmlFor="postal">Postal Code</Label>
    <Input placeholder="10179" id="postal" />
    <ButtonGroup align={ButtonGroup.LEFT}>
      <Button primary onClick={() => onNextClick()}>
        Submit
      </Button>
      <Button onClick={() => onBackClick()}>Back</Button>
    </ButtonGroup>
  </section>
);

const Thanks = () => (
  <section>
    <Heading>Thanks!</Heading>
  </section>
);

const MultiStepForm = () => {
  const steps = [FormOne, FormTwo, Thanks];
  const totalSteps = steps.length;

  return (
    <Step total={totalSteps}>
      {({ state, actions }) => {
        const StepComponent = steps[state.step];
        const stepNumber = state.step + 1;

        return (
          <div>
            <Heading size={Heading.GIGA}>
              Step {stepNumber} of {totalSteps}
            </Heading>
            <ProgressBar
              value={stepNumber}
              max={totalSteps}
              size={ProgressBar.KILO}
            />
            <StepComponent
              onNextClick={actions.next}
              onBackClick={actions.previous}
            />
          </div>
        );
      }}
    </Step>
  );
};
```

### Yes/No slider with [react-swipeable](https://github.com/dogfessional/react-swipeable)

And here is the interesting example of how you can easily add swipe/touch support into your slider-like UI if it's necessary:

![step-yes-no](https://user-images.githubusercontent.com/974035/65807605-846fc080-e18f-11e9-8642-e119434c7a2b.gif)
```js
import { Swipeable } from 'react-swipeable';
import { Step, Button, ButtonGroup } from '@sumup/circuit-ui';

const YesOrNoSlider = ({ images = [] }) => {
  const [swipe, setSwipe] = useState({});
  const handleSwipe = (eventData, actions) => {
    setSwipe(eventData);

    if (eventData.dir === LEFT_DIRECTION) {
      actions.previous();
    }

    if (eventData.dir === RIGHT_DIRECTION) {
      actions.next();
    }
  };

  return (
    <Step totalSteps={images.length}>
      {({ state, actions, getPreviousControlProps, getNextControlProps }) => (
        <div>
          <Swipeable onSwiped={eventData => handleSwipe(eventData, actions)}>
            <SliderImage
              src={images[state.step]}
              alt="random pic"
              swipe={swipe}
            />
          </Swipeable>
          <ButtonGroup align={ButtonGroup.CENTER}>
            <Button
              primary={swipe.dir === LEFT_DIRECTION}
              {...getPreviousControlProps()}
            >
              No
            </Button>
            <Button
              primary={swipe.dir === RIGHT_DIRECTION}
              {...getNextControlProps()}
            >
              Yes
            </Button>
          </ButtonGroup>
        </div>
      )}
    </Step>
  );
};
```

## Carousel component

`Carousel` is a port of component designed for the website and you can also see it in action as part of Testimonials component on [careers page](https://sumup.com/careers/) (more details in proposal - https://github.com/sumup/circuit-ui/issues/409). 

So currently there are 2 ways of using it.

### Stateful

First one is just as simple as passing array of images into it. All internal state management is handled by `Step` component automatically:

![carousel-stateful](https://user-images.githubusercontent.com/974035/65807380-6bb2db00-e18e-11e9-941b-6787678b1299.gif)
```js
import { Carousel } from '@sumup/circuit-ui';

const slides = [
  {  image: { src, alt } }, 
  {  image: { src, alt } }
  //...
];

<Carousel slides={slides} />
```

### Composed

Second approach is compositional which means you get the building blocks and arrange them in a way that you want. Keep in mind that you will need to manage state by yourself (or via `Step`):

![carousel-composed](https://user-images.githubusercontent.com/974035/65807483-f7c50280-e18e-11e9-84c2-a24b34898076.gif)
```js
import { CarouselComposed } from '@sumup/circuit-ui';

const { 
  Container, 
  Slides, 
  Slide, 
  SlideImage, 
  Controls, 
  Status,
  PrevButton,
  NextButton
} = CarouselComposed

const slides = [
  {  image: { src, alt } }, 
  {  image: { src, alt } }
  //...
];

const CustomCarousel = () => {
  const total = slides.length;
  const [step, setStep] = useState(0);
  const goBack = () => setStep(step === 0 ? total - 1 : step - 1);
  const goForward = () => setStep(step === total - 1 ? 0 : step + 1);

  return (
    <Container>
      <Slides>
        {slides.map(({ image }, index) => (
          <Slide
            key={index}
            index={index}
            step={step}
          >
            <SlideImage
              src={image.src}
              alt={image.alt}
            />
          </Slide>
        ))}
      </Slides>
      <Controls>
        <PrevButton onClick={goBack} />
        <Status step={step} total={total}  />
        <NextButton onClick={goForward} />
      </Controls>
    </Container>
  );
};
```

### Limitations

As I mentioned before `Step` is just purely state management abstraction and it doesn't contain any touch/swipe event support into it. I also decided not to introduce any 3rd party "swipeable" component dependency for now. This means that currently `Carousel` component doesn't support swiping of images on mobiles. I see 2 ways of solving it:

- obviously adding something like `react-swipeable` as a dependency to Circuit UI
- creating light `Swipe` component inside Circuit UI by ourselves (which should be addressed in a separate PR)

I leave it as an open question and your feedback is welcome.

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

connor-baer commented on src/components/Carousel/components/SlideImage/SlideImage.js:
```diff
@@ -0,0 +1,68 @@
+const imageStyles = css`
+  label: carousel__image;
+  img {
+    object-fit: cover;
```
`object-fit` is only supported in Edge 16+, so this would need to be polyfilled. This brings up an interesting question: what browsers do we want to support in v2?

voronianski replied:
This can be communicated in the docs regarding `Carousel` or any other component with latest features like "please include polyfill if you need support for old browsers".

As for browser support, I would vote for last 2 versions.

fernandofleury replied:
Does this mean then Edge and 11?

connor-baer replied:
I'd say only Edge. 

I mean, the only project that is required to support old browsers is the dashboard, because of the Android webviews. I don't think that one project should hold back Circuit UI. I like @voronianski's suggestion to mention the required polyfills, either on a component level or globally, the same way React does.

connor-baer replied:
I'd say only Edge. 

I mean, the only project that is required to support old browsers is the dashboard, because of the Android webviews. I don't think that one project should hold back Circuit UI. I like @voronianski's suggestion to mention the required polyfills, either on a component level or globally, the same way React does.

voronianski commented on src/components/AspectRatio/AspectRatio.js:
```diff
@@ -0,0 +1,96 @@
+  }
+);
+
+AspectRatio.displayName = 'AspectRatio';
```
yep, eslint was complaining :(

voronianski commented on src/components/Carousel/components/Progress/Progress.js:
```diff
@@ -0,0 +1,81 @@
+ * limitations under the License.
+ */
+
+import PropTypes from 'prop-types';
```
Unfortunately not yet, task is in the backlog (I think in "awaiting designer input" state)

voronianski commented on src/components/Carousel/components/SlideImage/SlideImage.js:
```diff
@@ -0,0 +1,68 @@
+  label: carousel__slideimage;
+  background: ${theme.colors.n100};
+`;
+// fix AspectRatio when upgrade to latest @emotion dependency
```
leftover comment from when I just started the branch with old emotion version ^^

voronianski commented on src/components/Step/examples/MultiStepForm.js:
```diff
@@ -0,0 +1,116 @@
+
+/* istanbul ignore file */
+
+import React from 'react';
```
thank youu!

voronianski commented on src/components/AspectRatio/AspectRatio.js:
```diff
@@ -0,0 +1,96 @@
+    const [child, ...restChildren] = Children.toArray(children);
+
+    return (
+      <Wrapper ref={ref} aspectRatio={aspectRatio} className={className}>
```
mm, but you can pass `ref` to child directly -

```js
const Example = () => {
  const parentRef = useRef(null);
  const childRef = useRef(null);

   return (
     <AspectRatio ref={parentRef}>
       <Image ref={childRef} />
     </AspectRatio>
   );
}
```

or am I missing something?

connor-baer replied:
You're totally right. I didn't think about that.

fernandofleury commented on src/components/Carousel/Carousel.spec.js:
```diff
@@ -0,0 +1,70 @@
+import Carousel from './Carousel';
+import { SLIDES } from './__fixtures__';
+
+describe('Carousel', () => {
```
Shouldn't we also add functional tests to the Carousel?

Like: 
- clicking the next slide and expecting it to be the active one
- clicking on the last slide and expecting it to loop
- waiting for the timer to change the active slide

voronianski replied:
yo @fernandofleury, thanks for suggestion, however all the points you mentioned are covered by the tests in useStep hook which is used inside Carousel component and is responsible for managing the state - https://github.com/sumup-oss/circuit-ui/blob/feature/add-carousel-and-step-components/src/components/Step/hooks/use-step.spec.js

Also all possible variations of internal Carousel component styles are tested separately inside components itself by passing necessarily props and matching snapshots (e.g. Slide component and different animations).

All components in Step and Carousel have 100% code coverage.

I agree that having functional/e2e test is nice addition but in current setup (react-testing-library) we will end up with just testing React onClick handlers calling the right functions. This is totally fine but it also feels like having too match hassle (e.g. adding a lot of data-selectors) for no real benefit.

WDYT?

fernandofleury commented on src/components/Carousel/components/Slide/Slide.js:
```diff
@@ -0,0 +1,160 @@
+
+const slideIn = keyframes`
+  from {
+    width: 0%;
```
Is it possible to use `transform: scaleX` instead of `width`? This way the animation can be optimised by the GPU.

voronianski replied:
it's possible but animation doesn't look the way it should be unfortunately.

![carousel-transform-scalex](https://user-images.githubusercontent.com/974035/73083032-bfc69580-3eca-11ea-8065-12a0bd5352e5.gif)

fernandofleury commented on src/components/Carousel/components/Slide/Slide.js:
```diff
@@ -0,0 +1,160 @@
+      animation-duration: ${animationDuration}ms;
+      animation-fill-mode: forwards;
+      animation-timing-function: ${theme.transitions.easeInOutCubic};
+    `};
```
```suggestion
      animation-name: ${animationName};
      animation-duration: ${animationDuration}ms;
      animation-fill-mode: forwards;
      animation-timing-function: ${theme.transitions.easeInOutCubic};
      animation-play-state: ${isAnimating ? 'running' : 'paused'}
```

Maybe we can use the play-state here? This way you don't have to conditionally add rules, just one value

voronianski replied:
Due to the way `animation-play-state` prop works, it leads to inconsistent sliding animations -
> Resuming a paused animation will start the animation from where it left off at the time it was paused, rather than starting over from the beginning of the animation sequence.

fernandofleury commented on src/components/Carousel/hooks/use-component-size.js:
```diff
@@ -0,0 +1,70 @@
+      };
+    }
+
+    window.addEventListener('resize', handleResize);
```
Shouldn't we throttle the resize listener?

fernandofleury commented on src/components/Step/StepService.js:
```diff
@@ -0,0 +1,95 @@
+  }
+}
+
+export function callAll(...fns) {
```
Since we already use lodash, can't we use something like [over](https://gist.github.com/jfmengels/6b973b69c491375117dc#_overiteratees)?

voronianski replied:
done

fernandofleury commented on src/components/Step/examples/CarouselSlider.js:
```diff
@@ -0,0 +1,109 @@
+        >
+          {images.map((image, i) => (
+            <SliderImage
+              key={i}
```
Can we use the image src or the alt for the key instead of the index? https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

fernandofleury replied:
Maybe it doesn't matter since this is an example.

voronianski replied:
fixed

voronianski commented on scripts/static-styles/index.js:
```diff
@@ -15,6 +15,7 @@
 
 import fs from 'fs';
 import path from 'path';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
added this line because eslint wasn't allowing me to commit changes -
```bash
18:1  error  'prettier' should be listed in the project's dependencies. Run 'npm i -S prettier' to add it  import/no-extraneous-dependencies
```

connor-baer commented:
@voronianski What's the state of this PR? Could you please address the review comments so we can merge it soon? Is there anything I can help you with? :)

voronianski commented:
> @voronianski What's the state of this PR? Could you please address the review comments so we can merge it soon? Is there anything I can help you with? :)

@connor-baer, thanks a lot but no help is needed, it's just a question of time :) will address this asap

voronianski commented:
> > @voronianski What's the state of this PR? Could you please address the review comments so we can merge it soon? Is there anything I can help you with? :)
> 
> @connor-baer, thanks a lot but no help is needed, it's just a question of time :) will address this asap

yo @connor-baer, I addressed all the comments, feel free to recheck the PR.

larimaza commented:
Can we close https://github.com/sumup-oss/circuit-ui/issues/409? @connor-baer
